esmvaltool/diag_scripts/iht_toa/poisson_solver.py

```diff
@@ -0,0 +1,281 @@
+"""(C) Crown Copyright 2023, the Met Office.
+
+Poisson solver for the full ocean-atmosphere column. The Poisson equation
+is solved by numerically using the biconjugate gradient stabilized (BiCGSTAB)
+method.
+
+The solution is achieved when the difference between the input field (radiative
+flux) and the Laplacian of the output field is less than the stated tolerance.
+If the solver fails to converge, the tolerance can be increased.
+
+Convergence is achieved faster by using a preconditioner on the output field.
+
+The heat transport is calculated as the gradient of the energy flux potential,
+the output of the Poisson solver.
+"""
+
+import sys
+import numpy as np
+from numba import jit
+
+
+def swap_bounds(fld):
+    """Extend the array by one in all directions.
+
+    As the array is periodic it allows for easier computations at
+    boundaries.
+    """
+    shp0, shp1 = np.array(fld.shape) - 2
+    wrap_pnt = int(shp1 / 2 + 1)
+    for i in range(1, shp1 + 1):
+        fld[0, i] = fld[1, wrap_pnt]
+        fld[shp0 + 1, i] = fld[shp0, wrap_pnt]
+        wrap_pnt += 1
+        if wrap_pnt > shp1:
+            wrap_pnt = 1
+
+    fld[:, 0] = fld[:, shp1]
+    fld[:, shp1 + 1] = fld[:, 1]
+
+    return fld
+
+
+def dot_prod(xxx, yyy):
+    """Calculate dot product of two matrices only over source term size."""
+    shp0, shp1 = np.array(xxx.shape) - 2
+    return (xxx[1:shp0 + 1, 1:shp1 + 1] * yyy[1:shp0 + 1, 1:shp1 + 1]).sum()
+
+
+@jit
+def precon(xxx, m_matrix):
+    """Preconditioner.
+
+    This is a wrapper to two steps that are optimised using jit.
+    """
+    cxxx = np.zeros(np.array(xxx.shape))
+    precon_a(xxx, m_matrix[1], m_matrix[2], m_matrix[4], cxxx)
+    cxxx = swap_bounds(cxxx)
+    precon_b(m_matrix[0], m_matrix[3], cxxx)
+    cxxx = swap_bounds(cxxx)
+    return cxxx
+
+
+@jit
+def precon_a(xxx, m_w, m_s, m_p, cxxx):
+    """First step of preconditioner."""
+    shp0, shp1 = np.array(cxxx.shape) - 2
+    for j in range(1, shp0 + 1):
+        for i in range(1, shp1 + 1):
+            cxxx[j, i] = m_p[j, i] * \
```

Comment:
Are the preconditioning steps also described by van der Vorst (1992)? If not, some explanation or another reference for the equations here would be useful.

Reply:
Yes, they are. I'm expanding the comments to make that clear.